File: backend/src/lib/http.ts
import type { APIGatewayProxyStructuredResultV2 } from 'aws-lambda';
import { ZodError, type ZodSchema } from 'zod';
import { badRequest, HttpError } from './errors.js';

const BASE_HEADERS: Record<string, string> = {
  'content-type': 'application/json; charset=utf-8',
  'cache-control': 'no-store',
  'x-content-type-options': 'nosniff',
};

export function json(
  statusCode: number,
  body: unknown,
  headers: Record<string, string> = {},
): APIGatewayProxyStructuredResultV2 {
  return {
    statusCode,
    headers: { ...BASE_HEADERS, ...headers },
    body: JSON.stringify(body),
  };
}

export const ok = (body: unknown) => json(200, body);
export const created = (body: unknown) => json(201, body);
export const noContent = (): APIGatewayProxyStructuredResultV2 => ({
  statusCode: 204,
  headers: BASE_HEADERS,
  body: '',
});

export function errorResponse(error: unknown): APIGatewayProxyStructuredResultV2 {
  if (error instanceof HttpError) {
    return json(error.status, {
      error: { code: error.code, message: error.message, details: error.details },
    });
  }

  if (error instanceof ZodError) {
    return json(400, {
      error: {
        code: 'VALIDATION_FAILED',
        message: 'Some fields need attention.',
        details: error.issues.map((issue) => ({
          field: issue.path.join('.'),
          message: issue.message,
        })),
      },
    });
  }

  console.error('unhandled_error', error);
  return json(500, {
    error: { code: 'INTERNAL_ERROR', message: 'Something went wrong on our side.' },
  });
}

export function parseBody<T>(schema: ZodSchema<T>, raw: string | undefined, isBase64 = false): T {
  if (!raw) throw badRequest('A request body is required');
  const text = isBase64 ? Buffer.from(raw, 'base64').toString('utf8') : raw;
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch {
    throw badRequest('Request body must be valid JSON');
  }
  return schema.parse(parsed);
}

File: backend/src/lib/http.py
import json

from lib.errors import HttpError, ValidationError, bad_request

BASE_HEADERS = {
    "content-type": "application/json; charset=utf-8",
    "cache-control": "no-store",
    "x-content-type-options": "nosniff",
}


def json_response(status_code, body, headers=None):
    return {
        "statusCode": status_code,
        "headers": {**BASE_HEADERS, **(headers or {})},
        "body": json.dumps(body, default=str),
    }


def ok(body):
    return json_response(200, body)


def created(body):
    return json_response(201, body)


def no_content():
    return {"statusCode": 204, "headers": BASE_HEADERS, "body": ""}


def error_response(error):
    if isinstance(error, HttpError):
        return json_response(
            error.status,
            {"error": {"code": error.code, "message": str(error), "details": error.details}},
        )
    if isinstance(error, ValidationError):
        return json_response(
            400,
            {
                "error": {
                    "code": "VALIDATION_FAILED",
                    "message": "Some fields need attention.",
                    "details": error.issues,
                }
            },
        )
    print("unhandled_error", error)
    return json_response(500, {"error": {"code": "INTERNAL_ERROR", "message": "Something went wrong on our side."}})


def parse_json_body(raw, is_base64=False):
    if not raw:
        raise bad_request("A request body is required")
    text = raw
    if is_base64:
        import base64

        text = base64.b64decode(raw).decode("utf-8")
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        raise bad_request("Request body must be valid JSON") from None


def parse_body(schema_fn, raw, is_base64=False):
    return schema_fn(parse_json_body(raw, is_base64))
